### ISSUE TITLE
OutOfMemoryError when using XMLUnit

### ISSUE THREAD

**Issue by laeubi** [xmlunit/xmlunit#256]
We see [OOM Errors when using XML-Unit](https://github.com/eclipse-tycho/tycho/issues/1864)
```
10:14:02  java.lang.OutOfMemoryError: Java heap space
10:14:02      at com.sun.org.apache.xerces.internal.dom.DeferredDocumentImpl.getNodeObject (DeferredDocumentImpl.java:940)
10:14:02      at com.sun.org.apache.xerces.internal.dom.DeferredElementNSImpl.synchronizeData (DeferredElementNSImpl.java:127)
10:14:02      at com.sun.org.apache.xerces.internal.dom.ElementNSImpl.getLocalName (ElementNSImpl.java:307)
10:14:02      at com.sun.org.apache.xerces.internal.dom.CoreDocumentImpl.importNode (CoreDocumentImpl.java:1545)
10:14:02      at com.sun.org.apache.xerces.internal.dom.CoreDocumentImpl.importNode (CoreDocumentImpl.java:1750)
10:14:02      at com.sun.org.apache.xerces.internal.dom.CoreDocumentImpl.importNode (CoreDocumentImpl.java:1750)
10:14:02      at com.sun.org.apache.xerces.internal.dom.CoreDocumentImpl.importNode (CoreDocumentImpl.java:1750)
10:14:02      at com.sun.org.apache.xerces.internal.dom.CoreDocumentImpl.cloneNode (CoreDocumentImpl.java:405)
10:14:02      at com.sun.org.apache.xerces.internal.dom.DocumentImpl.cloneNode (DocumentImpl.java:181)
10:14:02      at org.xmlunit.util.Nodes.stripWhitespace (Nodes.java:93)
10:14:02      at org.xmlunit.input.WhitespaceStrippedSource.<init> (WhitespaceStrippedSource.java:40)
10:14:02      at org.xmlunit.builder.DiffBuilder.wrap (DiffBuilder.java:441)
10:14:02      at org.xmlunit.builder.DiffBuilder.build (DiffBuilder.java:430)
```

Is there anything one must do to "clear" DiffBuilder instances so we probably acummulate memory?

**Comment by phbenisc** [comment]
The only time I have seen OOMs with XML Unit was with really large documents.

xmlunit uses the DOM API which means the whole document is loaded into the heap.

How large is document and the available heap?

Maybe it could be that the model is multiple times in memory while processing, but to my knowledge xmlunit does not remember the document in any way. That being said it is probably not a bad idea to loose the xmlunit instance between test runs. Some xml APIs are not threadsafe and a clean environment makes debugging easier.

`super(Nodes.stripWhitespace(Convert.toDocument(originalSource)));` 
is on line WhitespaceStrippedSource.java:40
so it could be that at that moment that you have 2 additional copies of the document in the memory/heap.

**Comment by laeubi** [comment]
@phbenisc many thanks, I have now profiled the application and at least I can verify that no instances related to `xmlunit` are retained/hard referenced somewhere so I assume you are right that it is either to large documents or to maybe processed at the same time!

**Comment by laeubi** [comment]
@phbenisc as it seems XML unit is currently using DOM API, I wonder if it not would be possible to [use StaX instead](https://docs.oracle.com/javase/tutorial/jaxp/stax/why.html)? That would be much less memory intensive.

**Comment by phbenisc** [comment]
@bodewig is the right person for feedback ;)

I guess at some point one could think that the memory problem might solve itself in the future, but somehow XML documents seem to grow faster then RAM sticks :O ...

So there are two potential problems with StaX: 

- xmlUnit is build upon DOM API so that would be basically mean reprogramming xmlunit (which practically is a hobby project of @bodewig )
- With StaX you cannot revisit Nodes, which also limits feature richeness.

If you have determined that memory is the problem you could try to avoid whitespace trimming/normalisation that could reduce the peak memory requirement. 
Basically every operation which would result in a new document. 

I suspect however that the memory saving is only a simple fraction of the originall memory requirement. 

Also giving the application more RAM could also work, but this is also not limitless.

**Comment by bodewig** [comment]
It never occurred to me `XPathContext` may be contributing to memory pressure this much, though. It's been a very long time since I implemented the meat of the difference engine, so I will need to re-remember why things are what they are. Maybe there is a way to slim down the memory usage - this may take some time.

**Comment by laeubi** [comment]
I think it would be fair for such purpose to have a limited set of features (currently we fall back to byte-by-byte comparison for large files), so for example different order of elements might be disabled in this regard.

If you like some example files you can take the content.jar here:
https://github.com/eclipse-equinox/p2/tree/master/bundles/org.eclipse.equinox.p2.tests/testData/2018-12

it looks innocent at first as it is only 3.5MB compressed, but the expanded xml is 45MB and results in a Diff of 250MB when compared to itself and I'm using the following configuration:

```
 Diff baselineDiff = DiffBuilder.compare(Input.fromByteArray(baselineBytes))//
                .withTest(Input.fromByteArray(reactorBytes))//
                .checkForSimilar()//
                .ignoreComments() //
                .ignoreWhitespace().build();
```

**Comment by bodewig** [comment]
Thank you @laeubi

The "limited set of features" would mean "no `ElementSelector` at all" which also means no way to get a SIMILAR result in most cases as swapped element order is the most common reason for SIMILAR vs EQUALS.

**Comment by laeubi** [comment]
> swapped element order is the most common reason for SIMILAR vs EQUALS

Is there somewhere an explanation what "similar" exactly means? The javadoc for example states that CDATA is something that is considered "similar" but ordering _might_ not be always similar (depending on the context, ex with xsd most of the time one defines a predefined order).

**Comment by bodewig** [comment]
In the end, the `DifferenceEvaluator` is the one who decides, the user guide https://github.com/xmlunit/user-guide/wiki/DifferenceEvaluator should be more comprehensive than the javadocs.

If your context does not allow elements to be swapped, then you'd typically specify a custom `DifferenceEvaluator` for example by [chaining](https://www.xmlunit.org/api/java/2.9.0/org/xmlunit/diff/DifferenceEvaluators.html#chain-org.xmlunit.diff.DifferenceEvaluator...-) an [`upgradeDifferencesToDifferent`](https://www.xmlunit.org/api/java/2.9.0/org/xmlunit/diff/DifferenceEvaluators.html#upgradeDifferencesToDifferent-org.xmlunit.diff.ComparisonType...-) to the [`Default`](https://www.xmlunit.org/api/java/2.9.0/org/xmlunit/diff/DifferenceEvaluators.html#Default)
